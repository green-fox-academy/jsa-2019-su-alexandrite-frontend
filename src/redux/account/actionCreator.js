import { LOCAL_URL } from 'react-native-dotenv';
import {
  LOGIN_USER_FAILURE,
  LOGIN_USER_SUCCESS,
  LOGOUT_SUCCESS,
} from './actionType';

const loginUserSuccess = (payload) => ({
  type: LOGIN_USER_SUCCESS,
  payload,
});

export const loginUserFail = (payload) => ({
  type: LOGIN_USER_FAILURE,
  payload,
});

const loginUser = (username, password) => (dispatch) => {
  const loginURL = `${LOCAL_URL}/users/login`;
  fetch(loginURL, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify({
      username,
      password,
    }),
  })
    .then((response) => {
      if (response.status === 200) {
        return response.json();
      }
      throw new Error('Oops! there is something wrong with your app');
    })
    .then((response) => {
      dispatch(loginUserSuccess(response));
    })
    .catch((error) => dispatch(loginUserFail(error)));
};

export const logOut = () => ({
  type: LOGOUT_SUCCESS,
});

export default loginUser;
