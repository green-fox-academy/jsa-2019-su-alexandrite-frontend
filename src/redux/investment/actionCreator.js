import { SERVER_URL, API_URL, API_KEY } from 'react-native-dotenv';
import {
  FETCH_PORTFOLIO_DETAILS_START,
  FETCH_PORTFOLIO_DETAILS_FAIL,
  FETCH_PORTFOLIO_DETAILS_SUCCESS,
} from './actionType';
import { round, moneyAmount2String } from '../../common/numbers';

export const fetchPortfolioDetailsStart = () => ({
  type: FETCH_PORTFOLIO_DETAILS_START,
});

export const fetchPortfolioDetailsFail = (payload) => ({
  type: FETCH_PORTFOLIO_DETAILS_FAIL,
  payload,
});

export const fetchPortfolioDetailsSuccess = (payload) => ({
  type: FETCH_PORTFOLIO_DETAILS_SUCCESS,
  payload,
});

export const calculatePortfolioValue = () => (dispatch, getState) => {
  const serverUrl = new URL(`${SERVER_URL}/investments/user/`);
  const headers = new Headers();
  const { accessToken } = getState().user;
  headers.append('authorization', `Bearer ${accessToken}`);
  dispatch(fetchPortfolioDetailsStart());
  fetch(serverUrl, { headers })
    .then((response) => {
      if (!response.ok) {
        switch (response.status) {
          case 401:
            throw new Error('Sorry, we cannot validate your identity. Please login and try again.');
          default:
            throw new Error('Oops, there\'s something wrong with our app.');
        }
      }
      return response.json();
    })
    .then((stocks) => {
      const symbols = stocks.map((stock) => stock.symbol);
      const apiUrl = new URL(`${API_URL}/stock/market/batch`);
      apiUrl.searchParams.append('symbols', symbols);
      apiUrl.searchParams.append('types', 'price,company,logo');
      apiUrl.searchParams.append('token', API_KEY);
      fetch(apiUrl)
        .then((res) => {
          if (!res.ok) {
            throw new Error('Oops, there\'s something wrong with our app.');
          }
          return res.json();
        })
        .then((res) => {
          const instruments = {};
          let totalValue = stocks
            .map(({ shares, symbol }) => {
              instruments[symbol] = {
                logo: res[symbol].logo.url,
                company: res[symbol].company.companyName,
                description: res[symbol].company.description,
                marketValue: round(shares * res[symbol].price),
              };
              return shares * res[symbol].price;
            })
            .reduce((a, b) => a + b);
          totalValue = moneyAmount2String((totalValue));
          dispatch(fetchPortfolioDetailsSuccess({
            totalValue,
            stocks,
            instruments,
          }));
        });
    })
    .catch((err) => dispatch(fetchPortfolioDetailsFail(err)));
};
