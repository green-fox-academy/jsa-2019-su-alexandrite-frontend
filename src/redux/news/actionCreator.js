import {
  SERVER_URL,
} from 'react-native-dotenv';
import {
  FETCH_INVESTED_STOCKS_START,
  FETCH_INVESTED_STOCKS_FAIL,
  FETCH_INVESTED_STOCKS_SUCCESS,
} from './actionType';

export const fetchInvestedStocksStart = () => ({
  type: FETCH_INVESTED_STOCKS_START,
});

export const fetchInvestedStocksFail = (payload) => ({
  type: FETCH_INVESTED_STOCKS_FAIL,
  payload,
});

export const fetchInvestedStocksSuccess = (payload) => ({
  type: FETCH_INVESTED_STOCKS_SUCCESS,
  payload,
});

export const fetchInvestedStocks = () => async (dispatch, getState) => {
  try {
    const serverUrl = new URL(`${SERVER_URL}/investments/user/`);

    const headers = new Headers();
    const { accessToken } = getState().user;
    headers.append('authorization', `Bearer ${accessToken}`);

    dispatch(fetchInvestedStocksStart());
    const investmentsResult = await fetch(serverUrl, { headers });
    if (!investmentsResult.ok) {
      switch (investmentsResult.status) {
        case 401:
          throw new Error('Sorry, we cannot validate your identity. Please login and try again.');
        default:
          throw new Error('Oops, there\'s something wrong with our app.');
      }
    }

    const stocks = await investmentsResult.json();
    const symbols = stocks.map(({ symbol }) => symbol);
    dispatch(fetchInvestedStocksSuccess(
      symbols,
    ));
  } catch (err) {
    dispatch(fetchInvestedStocksFail(err));
  }
};
