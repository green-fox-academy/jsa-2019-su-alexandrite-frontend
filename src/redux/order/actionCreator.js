import { SERVER_URL } from 'react-native-dotenv';

import {
  PURCHASE_STOCK_SUCCESS,
  PURCHASE_STOCK_FAIL,
} from './actionType';

const purchaseStockSuccess = (payload) => ({
  type: PURCHASE_STOCK_SUCCESS,
  payload,
});

const purchaseStockFail = (payload) => ({
  type: PURCHASE_STOCK_FAIL,
  error: payload,
});

const purchaseStock = (symbol, shares, type, accessToken) => (dispatch) => {
  const orderUrl = new URL(`${SERVER_URL}/order`);
  fetch(orderUrl, {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
      Authorization: `Bearer ${accessToken}`,
    },
    body: JSON.stringify(
      {
        symbol,
        shares,
        type,
      },
    ),
  })
    .then((response) => {
      if (response.status !== 201) {
        switch (response.status) {
          case 401:
            throw new Error('Sorry, we cannot validate your identity. Please login and try again.');
          default:
            throw new Error('Oops! there\'s something wrong with our app.');
        }
      }
      return response.json();
    })
    .then(() => dispatch(purchaseStockSuccess()))
    .catch((error) => dispatch(purchaseStockFail(error.message)));
};

export default purchaseStock;
