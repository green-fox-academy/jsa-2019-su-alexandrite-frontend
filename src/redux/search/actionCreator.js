import { API_KEY, API_URL } from 'react-native-dotenv';

import { SEARCH_START, SEARCH_SUCCESS, SEARCH_FAILED } from './actionType';

const searchStart = () => ({
  type: SEARCH_START,
});

const searchSuccess = (payload) => ({
  type: SEARCH_SUCCESS,
  payload,
});

const searchFailed = (payload) => ({
  type: SEARCH_FAILED,
  payload,
});

const searchStockData = (input) => (dispatch) => {
  const url = new URL(`${API_URL}/search/${input}`);
  url.searchParams.append('token', API_KEY);
  dispatch(searchStart());
  fetch(url)
    .then((response) => {
      if (!response.ok) {
        switch (response.status) {
          case 404:
            throw new Error(`Unknown symbols were provided when you search for ${input}.`);
          case 400:
            throw new Error(`Invalid values were supplied when search for ${input}.`);
          default:
            throw new Error('Oops, there\'s something wrong with our app.');
        }
      }
      return response.json();
    })
    .then((response) => dispatch(searchSuccess(response)))
    .catch((err) => dispatch(searchFailed(err)));
};

export default {
  searchStockData,
};
