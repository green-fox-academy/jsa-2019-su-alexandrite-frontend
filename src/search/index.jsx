import React from 'react';
import { View, Text } from 'react-native';
import SearchBar from './headerBar';

const navigationOptions = {
  header: () => <SearchBar />,
};

const search = () => (
  <View>
    <Text>Search Screen</Text>
  </View>
);

search.navigationOptions = navigationOptions;

export default search;
