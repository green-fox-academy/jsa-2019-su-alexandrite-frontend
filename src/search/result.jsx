import React from 'react';
import { useSelector } from 'react-redux';
import { View, Text, Image } from 'react-native';
import { TouchableHighlight } from 'react-native-gesture-handler';
import styles from './styles';

import tradeIcon from '../../assets/icons/search-page/trade.png';
import detailIcon from '../../assets/icons/search-page/detail.png';

export default () => {
  const isLoading = useSelector((state) => state.search.isLoading);
  const searchResult = useSelector((state) => state.search.result);
  const error = useSelector((state) => state.search.error);

  if (isLoading) {
    return <Text>Loading</Text>;
  }
  if (error) {
    return <Text>{error.message}</Text>;
  }
  return (
    <View>
      {searchResult.map((info) => (
        <View key={info.symbol} style={styles.resultList}>
          <View>
            <Text style={styles.stockName}>{info.symbol}</Text>
            <Text style={styles.stockExchange}>{info.exchange}</Text>
          </View>
          <View style={styles.resultButton}>
            <TouchableHighlight>
              <Image source={tradeIcon} />
            </TouchableHighlight>
            <TouchableHighlight style={styles.detailButton}>
              <Image source={detailIcon} />
            </TouchableHighlight>
          </View>
        </View>
      ))}
    </View>
  );
};
