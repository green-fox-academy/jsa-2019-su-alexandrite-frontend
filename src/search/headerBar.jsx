import React, { useState, useEffect } from 'react';
import { SafeAreaView, TextInput } from 'react-native';
import { useDispatch } from 'react-redux';
import { fetchData } from '../redux/search/actionCreater';

export default function SearchBar() {
  const [value, onChangeText] = useState('');
  const dispatch = useDispatch();
  useEffect(() => {
    let timer;
    if (value.length >= 2) {
      timer = setTimeout(() => {
        dispatch(fetchData(value));
        alert('times up');
      }, 3000);
    }
    return () => {
      if (timer) clearTimeout(timer);
    };
  }, [value]);

  function handleChange(text) {
    onChangeText(text);
    // clearTimeout(timer);
    // if (value.length >= 2) {
    //   timer = setTimeout(() => {
    //     dispatch(fetchData(value));
    //     alert('times up');
    //   }, 2000);
    // }
  }

  return (
    <SafeAreaView>
      <TextInput
        style={{ height: 40, backgroundColor: '#566ed3' }}
        onChangeText={(text) => handleChange(text)}
        value={value}
      />
    </SafeAreaView>
  );
}
