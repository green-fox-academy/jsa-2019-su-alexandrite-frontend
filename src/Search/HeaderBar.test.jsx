import React from 'react';
import { shallow } from 'enzyme';
import renderer from 'react-test-renderer';
import { useDispatch } from 'react-redux';
import { useNavigation } from 'react-navigation-hooks';
import HeaderBar from './HeaderBar';
import { searchStockData } from '../redux/search/actionCreator';

jest.mock('react-redux');
jest.mock('../redux/search/actionCreator');
jest.mock('react-navigation-hooks');
jest.mock('./SearchBarInput', () => '<SearchBarInput />');
describe('<HeaderBar />', () => {
  beforeEach(() => {
    useDispatch.mockReturnValue(() => { });
    useNavigation.mockReturnValue(() => ({
      pop: jest.fn(() => { }),
      push: jest.fn(() => { }),
    }));
    searchStockData.mockReturnValue('test_action');
  });

  afterAll(() => {
    jest.resetModules();
  });

  it('Expect to not log errors in console', () => {
    const spy = jest.spyOn(global.console, 'error');
    const wrapper = shallow(<HeaderBar />);
    expect(wrapper).not.toBeNull();
    expect(spy).not.toHaveBeenCalled();
  });

  it('Should render and match the snapshot', () => {
    const tree = renderer.create(
      <HeaderBar />,
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('Should call navigation.pop when cancel button is pressed', () => {
    const pop = jest.fn(() => { });
    const push = jest.fn(() => { });
    useNavigation.mockReturnValue({
      pop,
      push,
    });
    const wrapper = shallow(
      <HeaderBar />,
    );
    wrapper.find('TouchableHighlight')
      .first()
      .simulate('press');
    expect(useNavigation).toHaveBeenCalled();
    expect(pop).toHaveBeenCalled();
  });
});
