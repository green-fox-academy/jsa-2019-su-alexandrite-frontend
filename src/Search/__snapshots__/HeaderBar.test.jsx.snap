// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<HeaderBar /> Should render and match the snapshot 1`] = `
<RCTSafeAreaView
  emulateUnlessSupported={true}
  style={
    Object {
      "backgroundColor": "#566ed3",
    }
  }
>
  <View
    style={
      Object {
        "flexDirection": "row",
        "padding": 15,
      }
    }
  >
    <<SearchBarInput />
      setText={[Function]}
      text=""
    />
    <View
      accessible={true}
      focusable={true}
      isTVSelectable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "alignItems": "center",
          "justifyContent": "center",
          "marginLeft": 15,
        }
      }
    >
      <Text
        style={
          Object {
            "color": "#fff",
            "fontSize": 14,
            "fontWeight": "500",
          }
        }
      >
        Cancel
      </Text>
    </View>
  </View>
</RCTSafeAreaView>
`;
